packages/flutter/lib/src/material/refresh_indicator.dart

```diff
@@ -146,73 +165,80 @@ class _RefreshIndicatorState extends State<RefreshIndicator> {
     _maxScrollOffset = scrollBehavior.maxScrollOffset;
   }
 
-  void _onScrollStarted(ScrollableState scrollable) {
-    _updateState(scrollable);
-    _scaleController.value = 0.0;
-    _sizeController.value = 0.0;
-  }
-
   RefreshIndicatorLocation get _locationForScrollOffset {
     return _scrollOffset < _minScrollOffset
       ? RefreshIndicatorLocation.top
       : RefreshIndicatorLocation.bottom;
   }
 
-  void _onScrollUpdated(ScrollableState scrollable) {
+  void _handlePointerDown(PointerDownEvent event) {
+    final ScrollableState scrollable = config.scrollableKey?.currentState;
+    if (scrollable == null)
+      return;
+
+    _updateState(scrollable);
+    _scaleController.value = 0.0;
+    _sizeController.value = 0.0;
+    _mode = _RefreshIndicatorMode.drag;
+  }
+
+  void _handlePointerMove(PointerMoveEvent event) {
+    final ScrollableState scrollable = config.scrollableKey?.currentState;
+    if (scrollable == null)
+      return;
     final double value = scrollable.scrollOffset;
     if ((value < _minScrollOffset || value > _maxScrollOffset) &&
         ((value - _scrollOffset).abs() > kPixelScrollTolerance.distance)) {
       final double overScroll = value < _minScrollOffset ? _minScrollOffset - value : value - _maxScrollOffset;
       final double newValue = overScroll / (_containerExtent * _kDragContainerExtentPercentage);
-      if (newValue > _sizeController.value) {
-        _sizeController.value = newValue;
-        if (_location != _locationForScrollOffset) {
-          setState(() {
-            _location = _locationForScrollOffset;
-          });
-        }
+      _sizeController.value = newValue.clamp(0.0, 1.0);
+      if (_location != _locationForScrollOffset) {
+        setState(() {
+          _location = _locationForScrollOffset;
+        });
       }
     }
+    _mode = _valueColor.value.alpha == 0xFF ? _RefreshIndicatorMode.armed : _RefreshIndicatorMode.drag;
     _updateState(scrollable);
   }
 
-  Future<Null> _doOnScrollEnded(ScrollableState scrollable) async {
-    if (_valueColor.value.alpha == 0xFF) {
+  Future<Null> _doHandlePointerUp(PointerUpEvent event) async {
+    if (_mode == _RefreshIndicatorMode.armed) {
+      _mode = _RefreshIndicatorMode.snap;
       await _sizeController.animateTo(1.0 / _kDragSizeFactorLimit, duration: _kIndicatorSnapDuration);
-      await config.refresh();
-    }
-    return _scaleController.animateTo(1.0, duration: _kIndicatorScaleDuration);
-  }
+      if (_mode == _RefreshIndicatorMode.snap) {
+        setState(() {
+          _mode = _RefreshIndicatorMode.refresh; // Show the indeterminate progress indicator.
+        });
 
-  void _onScrollEnded(ScrollableState scrollable) {
-    _doOnScrollEnded(scrollable);
-  }
+        final Future<Null> refreshFuture = config.refresh();
+        _pendingRefreshFutures.insert(0, refreshFuture);
+        await refreshFuture;
+        _pendingRefreshFutures.remove(refreshFuture);
```

Comment:
As discussed, we might want to coalesce these updates for the client so that we only have one outstanding.  Rather than calling `config.refresh` unconditionally here, we can just check whether we have a pending future already and await it if we do.